lib/posix/clock.c

```diff
@@ -48,8 +48,8 @@ int z_impl_clock_gettime(clockid_t clock_id, struct timespec *ts)
 	}
 
 	uint64_t ticks = k_uptime_ticks();
-	uint64_t elapsed_secs = k_ticks_to_ms_floor64(ticks) / MSEC_PER_SEC;
-	uint64_t nremainder = ticks - k_ms_to_ticks_floor64(MSEC_PER_SEC * elapsed_secs);
+	uint64_t elapsed_secs = ticks / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+	uint64_t nremainder = ticks - elapsed_secs * CONFIG_SYS_CLOCK_TICKS_PER_SEC;
```

Comment:
This doesn't look right.  The conversion function should be precise and reduce to the division already on all systems where that's possible.  The bug is in the second line I think, where it feeds the result back into the reverse conversion; that's going to lose bits in configurations where the tick rate isn't divisible by ms.  Can you check that just the second line fixes the issue?

Also... I haven't pulled up the whole file to check, but I have a hard time believing that that nremainder value is anything but a footgun in an implementation of clock_gettime().  There shouldn't be any state held, and if we want higher precision values we should start there instead of back-converting from coarser units like seconds.

My notion of an idealized implementation here should be as simple as `uint64_t ns = k_ticks_to_us_floor64(k_uptime_ticks())`, and then just doing division/modulus ops to extract tv_nsec and tv_sec.

Reply:
@andyross - with just the second line, we get a failure of
```
START - test_clock_gettime_rollover
CONFIG_SYS_CLOCK_TICKS_PER_SEC: 100
rollover_s: 184467440737095516
t-1: {18446744073709551, 596000000}
```

Reply:
IMHO, the routines in time_units.h should have the set of `k_ticks_to_secs_*()` functions.  It's milli, micro, and nano, but not full seconds.  I guess the thought was that conversion to seconds is simple enough, one should write out the division, as is done here.  But as soon as you toss in rates determined at run time or non-integral clock speeds, it would be nice to not make everyone write their own code for it.

Reply:
> My notion of an idealized implementation here should be as simple as `uint64_t ns = k_ticks_to_us_floor64(k_uptime_ticks())`,

You meant `_to_ns` of course.  This expression does overflow before a 64-bit cycle counter overflows, for clock rates < 1 GHz.  It does take some 584 years to do this.  So realistically, it ought to be good enough for any reasonable value of a timeout or time delta or timestamp.

However, @cfriedt's test looks at values around where the cycle counter overflows, which is large enough to trigger the overflow.